File: btl_backend/models/Place.js
const mongoose = require('mongoose');

const placeSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true
  },
  address: {
    type: String,
    required: true,
    trim: true
  },
  latitude: {
    type: Number,
    required: true
  },
  longitude: {
    type: Number,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [{
    type: String,
    trim: true
  }],
  rating: {
    type: Number,
    min: 0,
    max: 5,
    default: 0
  },
  reviewCount: {
    type: Number,
    default: 0
  },
  category: {
    type: String,
    required: true,
    enum: ['coffee', 'travel', 'restaurant', 'attraction']
  },
  priceLevel: {
    type: Number,
    min: 1,
    max: 4,
    default: 2
  },
  features: [{
    type: String,
    enum: ['wifi', 'outdoor', 'parking', 'pet_friendly', 'wheelchair_accessible']
  }],
  openingHours: {
    type: String,
    trim: true
  },
  phone: {
    type: String,
    trim: true
  },
  website: {
    type: String,
    trim: true
  },
  reviews: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    rating: {
      type: Number,
      min: 1,
      max: 5,
      required: true
    },
    comment: {
      type: String,
      trim: true
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Update updatedAt on save
placeSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

// Calculate average rating
placeSchema.methods.calculateAverageRating = function() {
  if (this.reviews.length === 0) {
    this.rating = 0;
    this.reviewCount = 0;
    return;
  }

  const totalRating = this.reviews.reduce((sum, review) => sum + review.rating, 0);
  this.rating = totalRating / this.reviews.length;
  this.reviewCount = this.reviews.length;
};

module.exports = mongoose.model('Place', placeSchema);

File: btl_mobile/lib/models/place.dart
class Place {
  final String id;
  final String name;
  final String address;
  final double latitude;
  final double longitude;
  final String description;
  final List<String> images;
  final double rating;
  final int reviewCount;
  final String category; // coffee, travel, etc.
  final double priceLevel; // 1-4
  final List<String> features; // wifi, outdoor, etc.
  final String? openingHours;
  final String? phone;
  final String? website;

  Place({
    required this.id,
    required this.name,
    required this.address,
    required this.latitude,
    required this.longitude,
    required this.description,
    required this.images,
    required this.rating,
    required this.reviewCount,
    required this.category,
    required this.priceLevel,
    required this.features,
    this.openingHours,
    this.phone,
    this.website,
  });

  factory Place.fromJson(Map<String, dynamic> json) {
    return Place(
      id: json['id'],
      name: json['name'],
      address: json['address'],
      latitude: json['latitude'],
      longitude: json['longitude'],
      description: json['description'],
      images: List<String>.from(json['images']),
      rating: json['rating'].toDouble(),
      reviewCount: json['reviewCount'],
      category: json['category'],
      priceLevel: json['priceLevel'].toDouble(),
      features: List<String>.from(json['features']),
      openingHours: json['openingHours'],
      phone: json['phone'],
      website: json['website'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'address': address,
      'latitude': latitude,
      'longitude': longitude,
      'description': description,
      'images': images,
      'rating': rating,
      'reviewCount': reviewCount,
      'category': category,
      'priceLevel': priceLevel,
      'features': features,
      'openingHours': openingHours,
      'phone': phone,
      'website': website,
    };
  }
}
